cfpb/cfgov-refresh
# Form explainers in Wagtail

Adds the ability to create form explainers in Wagtail rather than as hard-coded pages in cfgov-refresh, like the existing [Loan Estimate](https://www.consumerfinance.gov/owning-a-home/loan-estimate/) and [Closing Disclosure](https://www.consumerfinance.gov/owning-a-home/closing-disclosure/) pages.

This PR focuses on enabling the Wagtail explainer functionality. Much of the front end code here is copied directly from the existing hard-coded explainers (including the Jinja template, CSS, and JS), with minor modifications to remove the tab functionality and accommodate Wagtail data instead of JSON. A future PR will add front end enhancements that can be evaluated against the baseline of these copied files.


## Additions

### New additions
- Form explainer app with
  - Form explainer Wagtail page type 
  - Form explainer Wagtail block
  - Form explainer Jinja page type template
- Custom CSS for the form explainer block admin display

### Copied files
- Existing [form explainer template](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/owning-a-home/_templates/form-explainer.html) copied into explainer app. Modifications include: 
  - removal of [tabs](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/owning-a-home/_templates/form-explainer.html#L23-L40)
  - accepts Wagtail data instead of imported JSON
  - adds numbered (`.image-map_overlay__category-1`) instead of named (`.image-map_overlay__checklist`) category classes
  - main template code is no longer wrapped in render macro
- Existing [form explainer less file](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/css/form-explainer.less) is copied into the `apps/form-explainer/css` directory. Modifications include:
  - removal of [tab styles](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/css/form-explainer.less#L211-L249)
  - numbered (`.image-map_overlay__category-1`) instead of named (`.image-map_overlay__checklist`) category classes
- Existing [FormExplainer js](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/js/form-explainer/form-explainer.js) is copied into the `apps/form-explainer/js` directory. Modifications include:
  - Removal of [placeholder](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/js/form-explainer/form-explainer.js#L524-L570) and tab code ([here](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/js/form-explainer/form-explainer.js#L627-L644) and [here](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/js/form-explainer/form-explainer.js#L136-L148))
- Form explainer init code is copied from [index.js](https://github.com/cfpb/cfgov-refresh/blob/f5bc72adf882914320d5f38e84280c763e40e105/cfgov/unprocessed/apps/owning-a-home/js/form-explainer/index.js) to on demand js directory

## Testing

1. Check out branch and run migration: `python cfgov/manage.py migrate` and `yarn gulp`
2. Create a new `Form Explainer` page
3. Add an `Explainer` block with image and notes & verify that content is output as expected. (You can use any image/note coordinates, but for verisimilitude here's the [first image](https://www.consumerfinance.gov/static/apps/owning-a-home/img/loan-estimate-H24B-1.639aeb7e50d5.png) from the Loan Estimate explainer and [a tool](https://www.zaneray.com/responsive-image-map/) that outputs image map coordinates)
4. Add a second category of notes and verify that its expandable displays below the first.
5. Add a second page and verify that pagination works as expected (image option: the [second Loan Estimate page](https://www.consumerfinance.gov/static/apps/owning-a-home/img/loan-estimate-H24B-2.294da31d7719.png))


## Screenshots
### Wagtail admin
<img width="1473" alt="Screen Shot 2020-02-04 at 4 11 48 PM" src="https://user-images.githubusercontent.com/778171/73876809-781a0500-480c-11ea-8750-48e58be9bbc9.png">

### Explainer preview
<img width="1219" alt="Screen Shot 2020-02-06 at 10 05 30 AM" src="https://user-images.githubusercontent.com/778171/73967939-baefe180-48cd-11ea-8e87-7766f0051a00.png">


## Todos (in future PR)
- Add CSS and JS updates
- Add no-js message
- Define limit on number of categories and add more category background colors
- Accessibility updates
- Improve Wagtail helper text

## Checklist


## Testing checklist

### Browsers


### Accessibility


### Other
